### PR TITLE
Mtreglia/cython setuptools module

### PR DESCRIPTION
**Pull request by mtreglia-gpsw**
# Version 0.3.0 
This Pr introduce the following upgrades: 

- Update README.md.
- Honor black.
- Refactored tests.
- Added new cpp_std flag to specify the c++ standard version.
- Refactored python package architecture to use a single version of
  vendor.py as setup for the the required modules to build the bindings.
- Removed CLI commands.
- Added the original setup __file__ argument to setup to pass the current
  working directory.
- Resolve config parser deprecations.
- Migration from setup.py to pyproject.toml.
- Added automatic git versioning system
- Enhance tox.ini tests.
- Added git-actions on pull request and push on master
  - Platform: [Unix, Darwin, Windows]
  - Python versions: [3.7, 3.8, 3.9, 3.10, 3.11]
- Removed Makefile.
- Removed travis.yml.

### CODE REVIEW

File: cython_setuptools/vendor.py [hunk 1]
```diff
@@ -7,6 +7,11 @@
 
 import setuptools
 
+# distutils is deprecated starting from python3.10
+# but the migration to setuptools is not completed
+# this import will change in the future
```
Comment (nsilvy-gpsw) [inline cython_setuptools/vendor.py:14]: You could add a FIXME mention to easily find it in the future?

File: pyproject.toml [hunk 2]
```diff
@@ -0,0 +1,44 @@
+[build-system]
+requires = ["setuptools==67.2.0", "wheel"]
+build-backend = "setuptools.build_meta"
+
+[project]
+name = "cython_setuptools"
+description = "Cython setuptools integration"
+dynamic = ["version"]
+readme = "README.rst"
+requires-python = "~=3.7"
```
Comment (rlavarec-gpsw) [inline pyproject.toml:14]: 3.10 ?

File: cython_setuptools/__init__.py [hunk 3]
```diff
@@ -1,3 +1,2 @@
-__version__ = "0.2.3"
-
-from .vendor import setup  # noqa
+from ._version import __version__
```
Comment (jbnoel-gpsw) [inline cython_setuptools/__init__.py:1]: Explain in commit message how it works.

File: cython_setuptools/vendor.py [hunk 4]
```diff
@@ -1,8 +1,9 @@
-import subprocess
+import argparse
+import configparser
 import os
 import os.path as op
 import shlex
-import argparse
+import subprocess
```
Comment (jbnoel-gpsw) [inline cython_setuptools/vendor.py]: This is not done by black.

File: cython_setuptools/vendor.py [hunk 5]
```diff
@@ -1,8 +1,9 @@
-import subprocess
+import argparse
+import configparser
```
Comment (jbnoel-gpsw) [inline cython_setuptools/vendor.py:2]: `configparser` import is introduced in this commit ?

**Review by jbnoel-gpsw — DISMISSED** [review]
This PR is really hard to review.

This commit is a mix of multiple things, to be split.
- tests: update according the new architecture
- tests: enhance tox tests platforms and python dist
- gitactions: test pipeline on push and pull requestn

Please see my comments.

File: setup.cfg [hunk 6]
```diff
@@ -1,36 +0,0 @@
...
-author_email = luper.rouch@gmail.com
-url = https://github.com/flupke/cython-setuptools
-classifiers =
-    Framework :: Setuptools Plugin
-    Topic :: Software Development :: Build Tools
-    License :: OSI Approved :: MIT License
-
-[options]
-packages = find:
-install_requires = click
```
[the remark this answers is not in the document]
Comment (mtreglia-gpsw) [reply setup.cfg:18]: It was the previous requirements, I'll remove it. Thanks

File: setup.cfg [hunk 7]
```diff
@@ -1,36 +0,0 @@
...
-[options.extras_require]
-dev =
-    pytest<4.0.0
-    virtualenv
-    pytest-virtualenv
-    six
-
-[options.entry_points]
-console_scripts =
-    cython-setuptools = cython_setuptools.cli:main
```
[the remark this answers is not in the document]
Comment (mtreglia-gpsw) [reply setup.cfg:32]: 😅 I'll clean it. Thanks

File: pyproject.toml (hunk 2 above)
Comment (mtreglia-gpsw) [reply pyproject.toml:14]: The python version requirements depends of the compatibility of the code and actually it works starting from python3.7.

File: pyproject.toml [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+[tool.setuptools.dynamic]
+version = { attr = "cython_setuptools.__version__" }
+
+[option]
+packages = "find:"
+include_package_data = "true"
+zip_safe = "false"
+
+[tool.isort]
+known_first_party = "cython_setuptools"
```
[the remark this answers is not in the document]
Comment (mtreglia-gpsw) [reply pyproject.toml:47]: Yes, I'll remove it. Thanks

File: cython_setuptools/vendor.py [hunk 9]
```diff
@@ -226,8 +226,8 @@ def parse_setup_cfg(fp, cythonize=False, pkg_config=None, base_dir=''):
     """
     if pkg_config is None:
         pkg_config = _run_pkg_config
-    config = configparser.SafeConfigParser()
-    config.readfp(fp)
+    config = configparser.ConfigParser()
+    config.read_file(fp)
```
[the remark this answers is not in the document]
Comment (mtreglia-gpsw) [reply cython_setuptools/vendor.py:228]: Added more explanation in the comment 👍

File: cython_setuptools/vendor.py [hunk 10]
```diff
@@ -5,13 +5,14 @@
...
-import setuptools
+os.environ["SETUPTOOLS_USE_DISTUTILS"] = "stdlib"
+
+import setuptools  # noqa
 
 # distutils is deprecated starting from python3.10
 # but the migration to setuptools is not completed
 # this import will change in the future
-from setuptools._distutils.ccompiler import get_default_compiler
-import configparser
```
[the remark this answers is not in the document]
Comment (mtreglia-gpsw) [reply cython_setuptools/vendor.py]: Added back. Thanks

File: cython_setuptools/vendor.py [hunk 11]
```diff
@@ -5,13 +5,14 @@
 import shlex
 import argparse
 
-import setuptools
+os.environ["SETUPTOOLS_USE_DISTUTILS"] = "stdlib"
+
+import setuptools  # noqa
```
[the remark this answers is not in the document]
Comment (mtreglia-gpsw) [reply cython_setuptools/vendor.py]: Added descriptions and links on the deprecation in the git comment.

File: README.md [hunk 12]
```diff
@@ -65,6 +64,7 @@ sources = foo.pyx
           bar.cpp
 include_dirs = eval(__import__('numpy').get_include())
 language = c++
+cpp_std = 11
```
[the remark this answers is not in the document]
Comment (mtreglia-gpsw) [reply README.md:67]: yes, thanks.

File: README.md [hunk 13]
```diff
@@ -27,15 +27,14 @@ First install the `cython-setuptools` vendor module in the package, next to
...
 ```shell
 $ cd your-python-project/
-$ cython-setuptools install
 ```
 
-Then use `cython-setuptools`' `setup()` in your `setup.py`:
+Then use `cython_setuptools`' `setup()` in your `setup.py`:
 
 ```python
-from cysetuptools import setup
```
[the remark this answers is not in the document]
Comment (mtreglia-gpsw) [reply README.md:36]: It's not actually just a rename, it's the architectural change. I'll add a better description in the commit message.

File: cython_setuptools/__init__.py (hunk 3 above)
Comment (mtreglia-gpsw) [reply cython_setuptools/__init__.py:1]: 👍

File: cython_setuptools/vendor.py (hunk 5 above)
Comment (mtreglia-gpsw) [reply cython_setuptools/vendor.py:2]: Resolved, Thanks

**Review by sfranz-gpsw — CHANGES_REQUESTED** [review]
Big work! Please look at my comments :)

File: tox.ini [hunk 14]
```diff
@@ -1,6 +1,21 @@
 [tox]
-envlist = py27,py37
```
Comment (jbnoel-gpsw) [inline tox.ini:2]: With drop Python 2 ?

File: cython_setuptools/__init__.py (hunk 3 above)
Comment (jbnoel-gpsw) [reply cython_setuptools/__init__.py:1]: How do we create a version ?

File: tox.ini (hunk 14 above)
Comment (mtreglia-gpsw) [reply tox.ini:2]: What is the interest of keeping it rather then the inability to use more recent python's features ?

File: tests/test_integration.py [hunk 15]
```diff
@@ -1,31 +1,38 @@
...
+    reason="Windows environment variable are not easy to manage",
+)
+def test_compile_and_run_external_cythonize_mode(virtualenv, tmp_dir):
+    _, pypkg_dir = _setup_source("setup-no-cythonize.py", tmp_dir)
+    virtualenv.run(f"pip install {cython_setuptools_path}")
+    virtualenv.run(f"export CYTHONIZE=1; pip install -e {pypkg_dir.resolve()}")
+    assert int(virtualenv.run("python -m bar", capture=True)) == 2
 
 
-@pytest.mark.slow
```
[the remark this answers is not in the document]
Comment (mtreglia-gpsw) [reply tests/test_integration.py:17]: the pytest marker is removed in a specific commit.

File: cython_setuptools/__init__.py (hunk 3 above)
Comment (mtreglia-gpsw) [reply cython_setuptools/__init__.py:1]: A specific version is created with a git tag.
After a target version, all the additional commits are taken in considerations in the `minor` version plus an hash: 

automatic version after an additional commit on tagged version `0.3.0`: 

` 0.3.1.dev1+gbc4cc18`

File: tests/conftest.py [hunk 16]
```diff
@@ -1,14 +1,22 @@
...
+
 import pytest
 
 
-@pytest.fixture
+@pytest.fixture()
 def tmp_dir(request):
     """
-    Return a :class:`py.path.local` object pointing to the "tmp" directory
+    Return a :class:`pathlib.Path` object pointing to the "tmp" directory
```
[the remark this answers is not in the document]
Comment (mtreglia-gpsw) [reply tests/conftest.py:12]: I fix it 👍  Thanks

File: tox.ini (hunk 14 above)
Comment (mtreglia-gpsw) [reply tox.ini:2]: Done 👍  Thanks

File: tests/test_integration.py [hunk 17]
```diff
@@ -1,22 +1,27 @@
+from pathlib import Path
 import os.path as op
 import shutil
 
 import pytest
+import cython_setuptools
 
 
 this_dir = op.dirname(__file__)
+cython_setuptools_path = (Path(cython_setuptools.__file__).parent).parent
```
Comment (jbnoel-gpsw) [inline tests/test_integration.py:9]: Stay consistent, use existing `op` in this commit
`cython_setuptools_path = op.dirname(this_dir)` (no need to import `cython_setuptools`)
or
`cython_setuptools_path = op.abspath(op.join(cython_setuptools.__file__, "..", ".."))`

If you really want to move to `pathlib`, it should be a dedicated commit changing the whole module.

File: cython_setuptools/vendor.py [hunk 18]
```diff
@@ -28,7 +28,7 @@ def setup(original_setup_file: str, cythonize: bool = True, **kwargs):
     Arguments:
         original_setup_file (str): it's the "__file__" of the original setup.py
         cythonize (bool): The *cythonize* argument controls the default mode of operation:
-                          set it to ``True`` if you don't distribute C files with your 
+                          set it to ``True`` if you don't distribute C files with your
```
Comment (jbnoel-gpsw) [inline cython_setuptools/vendor.py:29]: Please fix this extra space in original commit in this PR.

File: cython_setuptools/vendor.py [hunk 19]
```diff
@@ -14,11 +14,11 @@
 import configparser
 
 
-DEFAULTS_SECTION = 'cython-defaults'
-MODULE_SECTION_PREFIX = 'cython-module:'
-CYTHON_EXT = '.pyx'
-C_EXT = '.c'
-CPP_EXT = '.cpp'
+DEFAULTS_SECTION = "cython-defaults"
```
Comment (jbnoel-gpsw) [inline cython_setuptools/vendor.py:15]: You probably need to introduce the flake8/black config before this commit.
I think this change uses a custom config.

File: cython_setuptools/vendor.py [hunk 20]
```diff
@@ -1,17 +1,15 @@
-import subprocess
+import argparse
```
Comment (jbnoel-gpsw) [inline cython_setuptools/vendor.py:1]: introduce `isort` config before this commit ? may be no required since the default seems sufficient.

File: cython_setuptools/cli.py
Comment (jbnoel-gpsw) [inline cython_setuptools/cli.py:1]: Move this commit just after the architectural change.
Can dependency on `click` be removed in this commit ?

File: tests/test_integration.py [hunk 21]
```diff
@@ -24,13 +24,13 @@ def _setup_source(setup_name, tmp_dir):
 def test_compile_and_run_no_cythonize_mode(virtualenv, tmp_dir):
     setup_path, pypkg_dir = _setup_source('setup-no-cythonize.py', tmp_dir)
     virtualenv.run(f"pip install {cython_setuptools_path}")
-    virtualenv.run('CYTHONIZE=1 pip install -e %s' % pypkg_dir)
+    virtualenv.run(f"export CYTHONIZE=1; pip install -e {pypkg_dir.resolve()}")
```
Comment (jbnoel-gpsw) [inline tests/test_integration.py:29]: why `.resolve()` ? Explain in commit message ?

File: tests/test_integration.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (jbnoel-gpsw) [reply tests/test_integration.py:17]: How slow is this test ?

File: tests/test_integration.py [hunk 22]
```diff
@@ -18,6 +20,10 @@ def _setup_source(setup_name, tmp_dir: Path):
     return setup_path, pypkg_dir
 
 
+@pytest.mark.skipif(
+    platform.system() == "Windows",
+    reason="Windows env variable are not easy to manage.",
+)
```
Comment (jbnoel-gpsw) [inline tests/test_integration.py:25]: Robust cythonization is required on Windows as well...
See with @ncian-gpsw or @sdeker-gpsw if we can keep this test ?

File: tests/test_integration.py [hunk 23]
```diff
@@ -5,7 +5,6 @@
 
 import cython_setuptools
 
-
```
Comment (jbnoel-gpsw) [inline tests/test_integration.py]: This change is unrelated.
Black commit on the whole repo early in the PR ?

File: tox.ini [hunk 24]
```diff
@@ -1,6 +1,21 @@
...
+[gh-actions]
+python =
+    3.8: py38, 
+    3.9: py39, flake8
+    3.10: py310,
+    3.11: py311,
 
 [testenv]
 extras = dev
-commands = pytest {posargs}
```
Comment (jbnoel-gpsw) [inline tox.ini:6]: What about `{posargs}` ?

File: tox.ini [hunk 25]
```diff
@@ -1,6 +1,21 @@
...
-envlist = py37
+envlist = py37, py38, py39, py310, py311
+isolated_build = true
+
+[gh-actions]
+python =
+    3.8: py38, 
+    3.9: py39, flake8
+    3.10: py310,
+    3.11: py311,
```
Comment (jbnoel-gpsw) [inline tox.ini:10]: How GitHub actions are related to tox ?

File: .flake8 [hunk 26]
```diff
@@ -0,0 +1,4 @@
+[flake8]
+ignore = E128, E127, E126, W503, E203
+max-line-length = 140
+max-complexity = 30
```
Comment (jbnoel-gpsw) [inline .flake8:4]: Flake8 should come in its own commit earlier in the PR (with black changes)

**Review by jbnoel-gpsw — COMMENTED** [review]
Way easier to read ! See remaining comments.

File: tests/test_integration.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (mtreglia-gpsw) [reply tests/test_integration.py:17]: pytest -n 6 tests -> 6 passed in `14.97s`

pytest  tests -> 6 passed in `20.08s`

[the remark this answers is not in the document]
Comment (mtreglia-gpsw) [reply tests/test_integration.py:17]: but basically it's the same test as `test_compile_and_run_cythonize_mode ` but use the env variable instead of  setting cythonize to `True`  `setup(__file__, cythonize=True)`. 

I'm wondering if we really need to recompile all the bindings just to check the read of the env var CYTHONIZE ?

File: cython_setuptools/vendor.py (hunk 18 above)
Comment (mtreglia-gpsw) [reply cython_setuptools/vendor.py:29]: Fixed, Thanks

File: cython_setuptools/vendor.py (hunk 19 above)
Comment (mtreglia-gpsw) [reply cython_setuptools/vendor.py:15]: I'll move the aesthetics just at the end

File: tests/test_integration.py (hunk 17 above)
Comment (mtreglia-gpsw) [reply tests/test_integration.py:9]: use `op` 👍

File: tests/test_integration.py (hunk 21 above)
Comment (mtreglia-gpsw) [reply tests/test_integration.py:29]: removed with the usage of `op`

File: tox.ini (hunk 25 above)
Comment (mtreglia-gpsw) [reply tox.ini:10]: tox-gh-actions is a tox plugin which helps running tox on GitHub Actions with multiple different Python versions on multiple workers in parallel. This project is inspired by [tox-travis](https://github.com/tox-dev/tox-travis).

https://pypi.org/project/tox-gh-actions/

File: tox.ini (hunk 24 above)
Comment (mtreglia-gpsw) [reply tox.ini:6]: I think it was the, removed, slow marker

File: cython_setuptools/cli.py
Comment (mtreglia-gpsw) [reply cython_setuptools/cli.py:1]: 👍

File: cython_setuptools/vendor.py [hunk 27]
```diff
@@ -265,6 +274,12 @@ def _expand_cython_modules(config, cythonize, pkg_config, base_dir):
     return ret
 
 
+def _convert_cpp_std(version):
+    if get_default_compiler() == 'msvc':
```
[the remark this answers is not in the document]
Comment (sdeker-gpsw) [reply cython_setuptools/vendor.py]: (non-blocking)

File: setup.cfg [hunk 28]
```diff
@@ -5,6 +5,7 @@ description = Cython setuptools integration
 long_description = file: README.md
 long_description_content_type = text/markdown
 keywords = cython setuptools
+python_requires = '~=3.7'
```
Comment (sdeker-gpsw) [inline setup.cfg]: Just a question for me: What is meaning of `~=` here ? (Non blocking)

File: tests/test_integration.py (hunk 22 above)
Comment (sdeker-gpsw) [reply tests/test_integration.py:28]: we can make another exclusive test for windows with the following commands:
`virtualenv.run("$env:CYTHONIZE=1")`
`virtualenv.run(f"pip install -e {pypkg_dir}")
`virtualenv.run('$env:CYTHONIZE=""')`

File: tests/conftest.py [hunk 29]
```diff
@@ -1,14 +1,22 @@
...
     relative to the test file.
     """
-    tmp_dir = request.fspath.dirpath('tmp')
-    if tmp_dir.isdir():
-        tmp_dir.remove()
+    tmp_dir = request.fspath.dirpath(f"tmp-{uuid.uuid4()}")
     tmp_dir.mkdir()
-    return tmp_dir
+    yield tmp_dir
+
```
[the remark this answers is not in the document]
Comment (mtreglia-gpsw) [reply tests/conftest.py:18]: Write now it's working in windows and this is the thing. Do you have any alternative working suggestions  ?

File: setup.cfg (hunk 28 above)
Comment (mtreglia-gpsw) [reply setup.cfg]: It means support all python distribution `>= 3.7` and `< 4.x`

File: cython_setuptools/vendor.py (hunk 27 above)
[the remark this answers is not in the document]
Comment (mtreglia-gpsw) [reply cython_setuptools/vendor.py]: Yes why not, Thanks.

File: .flake8 (hunk 26 above)
Comment (jbnoel-gpsw) [reply .flake8:4]: Flake8 is an addition (not existing in setup.cfg)

File: pyproject.toml [hunk 30]
```diff
@@ -0,0 +1,45 @@
...
+
+[option]
+packages = "find:"
+include_package_data = "true"
+zip_safe = "false"
+
+[tool.isort]
+profile = "black"
+force_sort_within_sections = true
+line_length = 140
```
Comment (jbnoel-gpsw) [inline pyproject.toml:49]: isort is an addition (not existing in setup.cfg)

File: tests/conftest.py (hunk 29 above)
[the remark this answers is not in the document]
Comment (sdeker-gpsw) [reply tests/conftest.py:18]: Yess https://github.com/gopro/cython-setuptools/tree/sldkr/tmp_path

File: cython_setuptools/vendor.py [hunk 31]
```diff
@@ -4,15 +4,17 @@
 import sys
 import shlex
 import argparse
+import configparser
+
+os.environ["SETUPTOOLS_USE_DISTUTILS"] = "stdlib"
 
-import setuptools
+import setuptools  # noqa
```
[the remark this answers is not in the document]
Comment (mtreglia-gpsw) [reply cython_setuptools/vendor.py]: It was to set an env variable before the import without linters error. I removed, it's not required anymore.

File: tests/conftest.py (hunk 29 above)
[the remark this answers is not in the document]
Comment (mtreglia-gpsw) [reply tests/conftest.py]: I used the `tmp_path` default pytest fixture, thanks 👍

File: tests/pypkg/cysetuptools.py [hunk 32]
```diff
@@ -1 +0,0 @@
-../../cython_setuptools/vendor.py
```
[the remark this answers is not in the document]
Comment (sfranz-gpsw) [reply tests/pypkg/cysetuptools.py:1]: I see you removed it in a separate commit, thanks

**Comment by nsilvy-gpsw** [comment]
I'm not sure my review is relevant here, isn't it?

File: tests/conftest.py (hunk 29 above)
[the remark this answers is not in the document]
Comment (jbnoel-gpsw) [reply tests/conftest.py]: Good suggestion @sdeker-gpsw ! (Should we apply it to crafter as well ?)

**Review by jbnoel-gpsw — APPROVED** [review]
Thanks for this great and very clear PR.
After your rework, it was a pleasure to review it !
